zallet/src/components/json_rpc/methods/list_addresses.rs

```diff
@@ -0,0 +1,210 @@
+}
+
+impl AddressSource {
+    fn empty(source: &'static str) -> Self {
+        Self {
+            source,
+            transparent: None,
+            sapling: vec![],
+            unified: vec![],
+        }
+    }
+
+    fn has_data(&self) -> bool {
+        self.transparent.is_some() || !self.sapling.is_empty() || !self.unified.is_empty()
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct TransparentAddresses {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    addresses: Vec<String>,
+
+    #[serde(rename = "changeAddresses")]
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    change_addresses: Vec<String>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct SaplingAddresses {
+    #[serde(rename = "zip32KeyPath")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    zip32_key_path: Option<String>,
+
+    addresses: Vec<String>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct UnifiedAddresses {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    seedfp: Option<String>,
+
+    /// The ZIP 32 account index.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    account: Option<u32>,
+
+    addresses: Vec<UnifiedAddress>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub(crate) struct UnifiedAddress {
+    /// The diversifier index that the UA was derived at.
+    diversifier_index: u128,
+
+    /// The receiver types that the UA contains (valid values are "p2pkh", "sapling", "orchard").
+    receiver_types: Vec<String>,
+
+    /// The unified address corresponding to the diversifier.
+    address: String,
+}
+
+pub(crate) fn call(wallet: &WalletConnection) -> Response {
+    let mut imported_watchonly = AddressSource::empty("imported_watchonly");
+    let mut mnemonic_seed = AddressSource::empty("mnemonic_seed");
+
+    for account_id in wallet
+        .get_account_ids()
+        .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+    {
+        let account = wallet
+            .get_account(account_id)
+            .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+            // This would be a race condition between this and account deletion.
+            .ok_or(RpcErrorCode::InternalError)?;
+
+        let mut change_addresses = vec![];
+        for (taddr, metadata) in wallet
+            .get_transparent_receivers(account.id(), true)
+            .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?
+        {
+            if let Some(metadata) = metadata {
+                if metadata.scope() == TransparentKeyScope::INTERNAL {
+                    // Change addresses are never used in a UA receiver, so we need to
+                    // list them separately.
+                    change_addresses.push(taddr.encode(wallet.params()));
+                }
+            } else {
+                // TODO: Provide some way of determining whether this address is a UA receiver.
+            }
+        }
+
+        // TODO: Expose all addresses, not just the last generated.
```

Comment:
Add a `librustzcash` issue for the required `zcash_client_backend` functionality.

Reply:
https://github.com/zcash/librustzcash/issues/1740